torch/autograd/graph.py

```diff
@@ -439,27 +440,60 @@ def inner_hook(grad: torch.Tensor):
 
         return inner_hook
 
-    class Handle(RemovableHandle):
-        handles: Tuple[RemovableHandle, ...]
+    handles: List[RemovableHandle] = []
+    for i, t in enumerate(tensors):
+        handles.append(t.register_hook(get_inner_hook(i)))
 
-        def __init__(self, handles: Tuple[RemovableHandle, ...]):
-            self.handles = handles
+    return _MultiHandle(tuple(handles))
 
-        def remove(self):
-            for handle in self.handles:
-                handle.remove()
 
-        def __getstate__(self):
-            return self.handles
+def _register_any_hook(
+    tensors: Sequence[torch.Tensor],
+    fn: Callable[[Any], None],
+):
+    """
+    Register an "any" backward hook.
 
-        def __setstate__(self, state):
-            self.handles = state
+    The hook is called exactly once when the gradient with respect to *any* of
+    the tensors in :attr:`tensors` is computed and does not run thereafter. It
+    is a wrapper over :meth:`torch.Tensor.register_hook` that ensure that only
+    the first hook that fires actually runs.
 
-    handles: List[RemovableHandle] = []
-    for i, t in enumerate(tensors):
-        handles.append(t.register_hook(get_inner_hook(i)))
+    The hook should not modify its arguments.
+
+    This function returns a handle with a method ``handle.remove()`` that removes the hook.
+    """
+    ran_hook = False
+
+    @functools.wraps(fn)
+    def wrapped_fn(*args, **kwargs):
+        nonlocal ran_hook
+        if not ran_hook:
+            ran_hook = True
+            fn(*args, **kwargs)
+        return
+
+    handles = tuple(
+        tensor.register_hook(wrapped_fn) for tensor in tensors if tensor.requires_grad
+    )
+    return _MultiHandle(handles)
+
+
+class _MultiHandle(RemovableHandle):
```

Comment:
Previously, the `Handle` class was defined only in `register_multi_grad_hook()`. However, `Handle` was not public, so I kept this `_MultiHandle` private for now.